crates/wasmparser/src/validator/core.rs

```diff
@@ -0,0 +1,1106 @@
+            }
+            TypeRef::Memory(ty) => {
+                self.memories.push(ty);
+                (self.memories.len(), self.max_memories(features), "memories")
+            }
+            TypeRef::Tag(ty) => {
+                self.tags.push(ty);
+                (self.tags.len(), MAX_WASM_TAGS, "tags")
+            }
+            TypeRef::Global(ty) => {
+                self.globals.push(ty);
+                self.num_imported_globals += 1;
+                (self.globals.len(), MAX_WASM_GLOBALS, "globals")
+            }
+        };
+
+        check_max(len, 0, max, desc, offset)?;
+
+        // If the module is defined as part of a component, imports must be unique
+        if has_parent {
+            if self
+                .imports
+                .insert(
+                    (import.module.to_string(), import.name.to_string()),
+                    vec![entity],
+                )
+                .is_some()
+            {
+                return Err(BinaryReaderError::new(
+                    format!(
+                        "duplicate import name `{}:{}` already defined",
+                        import.module, import.name
+                    ),
+                    offset,
+                ));
+            }
+        } else {
+            self.imports
+                .entry((import.module.to_string(), import.name.to_string()))
+                .or_default()
+                .push(entity);
+        }
+
+        Ok(())
+    }
+
+    pub fn add_export(&mut self, export: &crate::Export, offset: usize) -> Result<()> {
+        let ty = self.export_to_entity_type(export, offset)?;
+        match self.exports.insert(export.name.to_string(), ty) {
+            Some(_) => Err(BinaryReaderError::new(
+                format!("duplicate export name `{}` already defined", export.name),
+                offset,
+            )),
+            None => Ok(()),
+        }
+    }
+
+    pub fn add_function(&mut self, type_index: u32, offset: usize) -> Result<()> {
+        self.types.func_type_at(type_index, offset)?;
+        self.functions.push(type_index);
+        Ok(())
+    }
+
+    pub fn add_table(
+        &mut self,
+        ty: TableType,
+        features: &WasmFeatures,
+        offset: usize,
+    ) -> Result<()> {
+        self.types.check_table_type(&ty, features, offset)?;
+        self.tables.push(ty);
+        Ok(())
+    }
+
+    pub fn add_memory(
+        &mut self,
+        ty: MemoryType,
+        features: &WasmFeatures,
+        offset: usize,
+    ) -> Result<()> {
+        self.types.check_memory_type(&ty, features, offset)?;
+        self.memories.push(ty);
+        Ok(())
+    }
+
+    pub fn add_global(
+        &mut self,
+        global: Global,
+        features: &WasmFeatures,
+        offset: usize,
+    ) -> Result<()> {
+        self.types.check_global_type(&global.ty, features, offset)?;
+        self.function_references.extend(self.check_init_expr(
```

Comment:
I've moved `check_init_expr` up to `ModuleState` so it can `self.module.assert_mut` where appropriate (and also check `self.order` rather than being told it's in the data section), so this simplifies things a bit.